From 6a019795908ac4b65497d7c5129b98106ff02f30 Mon Sep 17 00:00:00 2001
From: Kishore Kadiyala <kishore.kadiyala@intel.com>
Date: Thu, 1 Oct 2020 01:39:11 +0530
Subject: [PATCH 60/61] Misc: Fix

---
 fs/f2fs/data.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/fs/f2fs/data.c b/fs/f2fs/data.c
index 6e9017e6a819..61bc2c5e95d1 100644
--- a/fs/f2fs/data.c
+++ b/fs/f2fs/data.c
@@ -2875,8 +2875,7 @@ static int f2fs_write_cache_pages(struct address_space *mapping,
 		clear_inode_flag(mapping->host, FI_HOT_DATA);
 
 	if (wbc->range_cyclic) {
-		writeback_index = mapping->writeback_index; /* prev offset */
-		index = writeback_index;
+		index = mapping->writeback_index; /* prev offset */
 		end = -1;
 	} else {
 		index = wbc->range_start >> PAGE_SHIFT;
-- 
2.25.1

